qiskit_experiments/data_processing/nodes.py

```diff
@@ -411,6 +433,135 @@ def _process(self, data: np.array) -> np.array:
+    a list of lists and returns a list of labels as do most Sklearn classifiers. Crucially,
+    this node can be initialized with a single discriminator which applies to each memory
+    slot or it can be initialized with a list of discriminators, i.e. one for each slot.
+    """
+
+    def __init__(
+        self, discriminator: Union[Any, List[Any]], validate: bool = True, no_ufloat: bool = True
+    ):
+        """Initialize the node with an object that can discriminate.
+
+        Args:
+            discriminator: The entity that will perform the discrimination. This needs to
+                be an object or a list of objects with a :meth:`predict` method that takes
+                as input a list of lists and returns a list of labels. If a list of
+                discriminators is given then there should be as many discriminators as there
+                will be slots in the memory.
+            no_ufloat: A boolean which if set to True, i.e., the default, will cause the node
+                to convert the data from uncertainties to regular floats.
+        """
+        super().__init__(1.0, validate)
+        self._discriminator = discriminator
+        self._no_ufloat = no_ufloat
+
+    def _format_data(self, data: np.ndarray) -> np.ndarray:
+        """Check that there are as many discriminators as there are slots."""
+        data = super()._format_data(data)
+
+        if self._validate:
+            if isinstance(self._discriminator, list):
+                if self._n_slots != len(self._discriminator):
+                    raise DataProcessorError(
+                        f"The Discriminator node has {len(self._discriminator)} which does "
+                        f"not match the {self._n_slots} slots in the data."
+                    )
+
+        if self._no_ufloat:
+            data = unp.nominal_values(data)
+
+        return data
+
+    def _process(self, data: np.ndarray) -> np.ndarray:
+        """Discriminate the data.
+
+        Args:
+            data: The IQ data as a list of points to discriminate. This data should have
+                the shape dim_1 x dim_2 x ... x dim_k x 2.
+
+        Returns:
+            The discriminated data as a list of labels with shape dim_1 x ... x dim_k.
+        """
+        # Case where one discriminator is applied to all the data.
+        if not isinstance(self._discriminator, list):
+            # Reshape the IQ data to an array of size n x 2
+            shape, data_length = data.shape, 1
+            for dim in shape[0:-1]:
+                data_length *= dim
+
+            data = data.reshape((data_length, 2))  # the last dim is guaranteed by _process
+
+            # Classify the data using the discriminator and reshape it to dim_1 x ... x dim_k
+            classified = np.array(self._discriminator.predict(data)).reshape(shape[0:-1])
+
+        # case where a discriminator is applied to each slot.
+        else:
+            if self._n_shots == 0:
```

Comment:
Does it make sense to run discriminator on averaged IQ? I think this should fall into `DataProcessorError`

Reply:
Mathematically you can do it which does not mean that it makes sense to do it. I think we can come up with some strange cases, e.g., use a linear regression as a discriminator to go straight to probability but that does not match the intended use case here. I'm tempted to raise a `DataProcessorError` as you suggest.

Reply:
See here: [61db3a5](https://github.com/Qiskit/qiskit-experiments/pull/805/commits/61db3a52f2374df4b9134d86c17aab5a47b4017d)